cli/support.go

```diff
@@ -0,0 +1,202 @@
+				}
+				tsStr := time.Now().In(loc).Format("2006-01-02-150405")
+				fname := "coder-support-" + tsStr + ".zip"
+				outputPath = filepath.Join(cwd, fname)
+			}
+
+			w, err := os.Create(outputPath)
+			if err != nil {
+				return xerrors.Errorf("create output file: %w", err)
+			}
+			zwr := zip.NewWriter(w)
+			defer zwr.Close()
+			if err := writeBundle(bun, zwr); err != nil {
+				return xerrors.Errorf("write support bundle to %s: %w", outputPath, err)
+			}
+			return nil
+		},
+		Hidden: true, // TODO: un-hide
+	}
+	cmd.Options = clibase.OptionSet{
+		{
+			Flag:          "output",
+			FlagShorthand: "o",
+			Env:           "CODER_SUPPORT_BUNDLE_OUTPUT",
+			Description:   "Path to which to output the generated support bundle. Defaults to coder-support-YYYYmmdd-HHMMSS.zip.",
+			Value:         clibase.StringOf(&outputPath),
+		},
+	}
+
+	return cmd
+}
+
+func findAgent(agentName string, haystack []codersdk.WorkspaceResource) (*codersdk.WorkspaceAgent, bool) {
+	for _, res := range haystack {
+		for _, agt := range res.Agents {
+			if agentName == "" {
+				// just return the first
+				return &agt, true
+			}
+			if agt.Name == agentName {
+				return &agt, true
+			}
+		}
+	}
+	return nil, false
+}
+
+func writeBundle(src *support.Bundle, dest *zip.Writer) error {
+	for k, v := range map[string]any{
+		"deployment/buildinfo.json":    src.Deployment.BuildInfo,
+		"deployment/config.json":       src.Deployment.Config,
+		"deployment/experiments.json":  src.Deployment.Experiments,
+		"deployment/health.json":       src.Deployment.HealthReport,
+		"network/netcheck_local.json":  src.Network.NetcheckLocal,
+		"network/netcheck_remote.json": src.Network.NetcheckRemote,
+		"workspace/workspace.json":     src.Workspace.Workspace,
+		"workspace/agent.json":         src.Workspace.Agent,
+	} {
+		f, err := dest.Create(k)
+		if err != nil {
+			return xerrors.Errorf("create file %q in archive: %w", k, err)
+		}
+		enc := json.NewEncoder(f)
+		enc.SetIndent("", "    ")
+		if err := enc.Encode(v); err != nil {
+			return xerrors.Errorf("write json to %q: %w", k, err)
+		}
+	}
+
+	for k, v := range map[string]string{
+		"network/coordinator_debug.html":   src.Network.CoordinatorDebug,
+		"network/tailnet_debug.html":       src.Network.TailnetDebug,
+		"workspace/build_logs.txt":         humanizeBuildLogs(src.Workspace.BuildLogs),
+		"workspace/agent_startup_logs.txt": humanizeAgentLogs(src.Workspace.AgentStartupLogs),
+		"logs.txt":                         strings.Join(src.Logs, "\n"),
+	} {
+		f, err := dest.Create(k)
+		if err != nil {
+			return xerrors.Errorf("create file %q in archive: %w", k, err)
+		}
+		if _, err := f.Write([]byte(v)); err != nil {
+			return xerrors.Errorf("write file %q in archive: %w", k, err)
+		}
+	}
+	if err := dest.Close(); err != nil {
+		return xerrors.Errorf("close zip file: %w", err)
+	}
+	return nil
+}
+
+func humanizeAgentLogs(ls []codersdk.WorkspaceAgentLog) string {
+	var l log.Logger
+
+	var buf bytes.Buffer
```

Comment:
Why do we use an intermediary vs passing in the writer? Perhaps can be resolved by wrapping in funcs.

Reply:
Tried it, makes things more complicated than I'd like.